### PR TITLE
[diffusion] Qwen-Image perf: mul add fusion

### PR DESCRIPTION
**Pull request by yingluosanqian**
## Motivation

Speed up Qwen-Image through kernel fusion.

I carefully analyzed the performance bottlenecks of Qwen Image, focusing on the MMDiT block, and drew out its structural decomposition. Based on this analysis, I found that the current implementation still has room for optimization. In particular, I plan to fuse as many memory-bound kernels as possible (highlighted as blue rectangles in the diagram).

As below figure shows, I group the computation into three large fused operators. The QwenImageEdit MMDit block is slightly different: its image branch introduces an additional modulated index, which requires conditional selection based on a 0/1 mask. As a result, it needs extra fused kernels to handle this logic.

image
image

My plan consists of the following three parts:
1. norm_scale_shift and scale_residual_norm_scale_shift (implemented in [this PR](https://github.com/sgl-project/sglang/pull/14717))

2. norm_scale_shift_select01 and scale_residual_norm_scale_shift_select01 (TODO. not in this PR)

3. scale_shift (this PR)

## Modifications

To avoid duplicating kernels, this implementation modifies the existing triton kernel to support a more general form, namely `a * b + c`. The original kernel implemented x * (1 + scale) + shift, and we extend it to handle the generalized case x * (0 + scale) + shift.

`scale_constant` is marked as a compile-time value since it only takes 0 or 1, avoiding runtime perf loss.

## Accuracy Tests

E2E test

`sglang generate --model-path Qwen/Qwen-Image --prompt "A Logo With Bold Large Text: SGL Diffusion" --seed 0`

image

## Benchmarking and Profiling

Before fusing: 55 us per forward
image

After fusing: 20 us per forward
image

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

**Review by yhyang201 — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py [hunk 1]
```diff
@@ -710,7 +713,7 @@ def _modulate(self, x, mod_params, index=None):
                 )
                 gate_result = torch.where(mask, gate0.unsqueeze(1), gate1.unsqueeze(1))
                 return (
-                    fuse_scale_shift_kernel(x, scale_result, shift_result),
+                    x * (1 + scale_result) + shift_result,
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py]: Why not use fuse_kernel here?

Comment (yingluosanqian) [reply python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py]: I remember the triton kernel (fused_scale_shift) does not support the CPU backend. Since this code path is executed when is_cuda is false, it should not be calling the Triton kernel.

Here someone implements such logic:
``` python
if is_cuda:
  ...
  fuse_scale_shift_gate_select01_kernel
else:
  ...
  return x * (1 + scale_result) + shift_result
```
IIUC, we should avoid using `_is_cuda` in the model definition. It would be better to encapsulate this logic in a `CustomOp` and let it handle dispatch automatically.
I will fix the `x * (1 + scale_result) + shift_result` at once.
Also, Ruby is currently implementing a fusion kernel for `fuse_scale_shift_gate_select01_kernel`, so she can take care of removing or fixing the `is_cuda` logic there.

Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py]: Ok

**Review by mickqian — APPROVED** [review]

**Comment by mickqian** [comment]
could you attach the performance report mentioned in  https://github.com/sgl-project/sglang/blob/main/python/sglang/multimodal_gen/docs/contributing.md#3-ci-based-change-protection?

**Comment by yingluosanqian** [comment]
The perf gain is small, as this is a very small fusion, yielding approximately 0.2%, 0.3%, and 0.2% speedups in the denoising stage for Qwen, Wan, and HunyuanVideo, respectively.

### Performance Comparison Report (Qwen)

#### 1. High-level Summary
| Metric | Baseline | New | Diff | Status |
| :--- | :--- | :--- | :--- | :--- |
| **E2E Latency** | 14043.03 ms | 14014.70 ms | **-28.33 ms (-0.2%)** | ⚪️ |
| **Throughput** | 0.07 req/s | 0.07 req/s | - | - |


#### 2. Stage Breakdown
| Stage Name | Baseline (ms) | New (ms) | Diff (ms) | Diff (%) | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| InputValidationStage | 0.04 | 0.04 | -0.00 | -4.9% | ⚪️ |
| TextEncodingStage | 39.21 | 39.55 | +0.34 | +0.9% | ⚪️ |
| ConditioningStage | 0.01 | 0.01 | -0.00 | -5.6% | ⚪️ |
| TimestepPreparationStage | 0.49 | 0.47 | -0.02 | -4.6% | ⚪️ |
| LatentPreparationStage | 0.14 | 0.13 | -0.01 | -4.5% | ⚪️ |
| DenoisingStage | 13938.76 | 13912.40 | -26.36 | -0.2% | ⚪️ |
| DecodingStage | 62.62 | 60.50 | -2.11 | -3.4% | ⚪️ |


<details>
<summary>Metadata</summary>

- Baseline Commit: `c0b5a180febef52bba4f1f3e974d364f6c96a10f`
- New Commit: `e64acd2c1b823a5b9b637fa3254358a2e45d26bc`
- Timestamp: 2026-01-28T04:34:27.148156
</details>

### Performance Comparison Report (Wan)

#### 1. High-level Summary
| Metric | Baseline | New | Diff | Status |
| :--- | :--- | :--- | :--- | :--- |
| **E2E Latency** | 357825.13 ms | 356981.45 ms | **-843.68 ms (-0.2%)** | ⚪️ |
| **Throughput** | 0.00 req/s | 0.00 req/s | - | - |


#### 2. Stage Breakdown
| Stage Name | Baseline (ms) | New (ms) | Diff (ms) | Diff (%) | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| InputValidationStage | 0.06 | 0.06 | +0.01 | +12.4% | ⚪️ |
| TextEncodingStage | 1112.50 | 1141.99 | +29.50 | +2.7% | ⚪️ |
| ConditioningStage | 0.02 | 0.02 | +0.00 | +16.6% | ⚪️ |
| TimestepPreparationStage | 0.57 | 0.57 | -0.01 | -1.0% | ⚪️ |
| LatentPreparationStage | 0.21 | 0.19 | -0.02 | -10.7% | ⚪️ |
| DenoisingStage | 347787.02 | 346881.44 | -905.58 | -0.3% | ⚪️ |
| DecodingStage | 8193.28 | 8224.68 | +31.40 | +0.4% | ⚪️ |


<details>
<summary>Metadata</summary>

- Baseline Commit: `c0b5a180febef52bba4f1f3e974d364f6c96a10f`
- New Commit: `e64acd2c1b823a5b9b637fa3254358a2e45d26bc`
- Timestamp: 2026-01-28T04:34:27.207503
</details>

### Performance Comparison Report (Hunyuan)

#### 1. High-level Summary
| Metric | Baseline | New | Diff | Status |
| :--- | :--- | :--- | :--- | :--- |
| **E2E Latency** | 71013.10 ms | 70715.54 ms | **-297.57 ms (-0.4%)** | ⚪️ |
| **Throughput** | 0.01 req/s | 0.01 req/s | - | - |


#### 2. Stage Breakdown
| Stage Name | Baseline (ms) | New (ms) | Diff (ms) | Diff (%) | Status |
| :--- | :--- | :--- | :--- | :--- | :--- |
| InputValidationStage | 0.05 | 0.04 | -0.01 | -26.6% | ⚪️ |
| TextEncodingStage | 309.98 | 305.08 | -4.90 | -1.6% | ⚪️ |
| ConditioningStage | 0.01 | 0.01 | -0.01 | -39.1% | ⚪️ |
| TimestepPreparationStage | 0.47 | 0.36 | -0.10 | -22.4% | ⚪️ |
| LatentPreparationStage | 0.16 | 0.11 | -0.05 | -32.4% | ⚪️ |
| DenoisingStage | 53499.09 | 53412.46 | -86.63 | -0.2% | ⚪️ |
| DecodingStage | 17200.98 | 16995.56 | -205.41 | -1.2% | ⚪️ |


<details>
<summary>Metadata</summary>

- Baseline Commit: `c0b5a180febef52bba4f1f3e974d364f6c96a10f`
- New Commit: `e64acd2c1b823a5b9b637fa3254358a2e45d26bc`
- Timestamp: 2026-01-28T04:34:27.266569
</details>
